File: seqeron-client/src/test/cpp/org/limitless/seqeron/app/GatewayLifecycleTest.cpp
// Unit tests for the gateway election lifecycle, merged from phixeron's acceptor and initiator tests.
// Topology: GW-A (id 5) and GW-B (id 6) under gatewaySourceId 6, beside another pair (ids 1 and 2) under 0.
// Case for case with GatewayLifecycleTest.java.

#include <algorithm>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "org/limitless/seqeron/app/GatewayLifecycle.hpp"

namespace org::limitless::seqeron::app {
namespace {

constexpr const char* ME = "GW-A";
constexpr std::int32_t MY_ID = 5;
constexpr std::int32_t SIBLING_ID = 6;
constexpr std::int32_t MY_SOURCE_ID = 6;
constexpr std::int32_t OTHER_PAIR_ID = 1;
constexpr std::int32_t OTHER_PAIR_SOURCE_ID = 0;

struct RecordingActions
{
    void identityResolved(const std::int32_t gatewayId, const std::int32_t gatewaySourceId, std::int32_t)
    {
        calls.push_back("identity(" + std::to_string(gatewayId) + "," + std::to_string(gatewaySourceId) + ")");
    }

    bool publishGatewayStarted(const std::int32_t gatewayId)
    {
        calls.push_back("GatewayStarted(" + std::to_string(gatewayId) + ")");
        return gatewayStartedLands;
    }

    bool openGate()
    {
        calls.emplace_back("open");
        return gateOpens;
    }

    void closeGate()
    {
        calls.emplace_back("close");
    }

    [[nodiscard]] long count(const std::string& call) const
    {
        return std::ranges::count(calls, call);
    }

    std::vector<std::string> calls;
    bool gatewayStartedLands = true;
    bool gateOpens = true;
};

using Lifecycle = GatewayLifecycle<RecordingActions>;
using State = Lifecycle::State;

const std::string MY_STARTED = "GatewayStarted(" + std::to_string(MY_ID) + ")";

class GatewayLifecycleTest : public testing::Test
{
  protected:
    void loadTopology()
    {
        lifecycle.onGatewayRegistered(OTHER_PAIR_ID, OTHER_PAIR_SOURCE_ID, "OTHER-A", 0);
        lifecycle.onGatewayRegistered(2, OTHER_PAIR_SOURCE_ID, "OTHER-B", 1);
        lifecycle.onGatewayRegistered(MY_ID, MY_SOURCE_ID, ME, 0);
        lifecycle.onGatewayRegistered(SIBLING_ID, MY_SOURCE_ID, "GW-B", 1);
    }

    void becomeServing()
    {
        loadTopology();
        lifecycle.onGatewayActive(MY_ID);
        lifecycle.onCaughtUp();
        lifecycle.advance();
    }

    RecordingActions actions;
    Lifecycle lifecycle{ ME, actions };
};

TEST_F(GatewayLifecycleTest, ResolvesItsIdentityFromTheRowNamingIt)
{
    loadTopology();
    EXPECT_EQ(MY_ID, lifecycle.gatewayId());
    EXPECT_EQ(MY_SOURCE_ID, lifecycle.gatewaySourceId());
    EXPECT_EQ(std::vector<std::string>{ "identity(5,6)" }, actions.calls);
}

TEST_F(GatewayLifecycleTest, IgnoresAnActivationBeforeAnyRowNamesThisInstance)
{
    lifecycle.onGatewayActive(MY_ID);
    loadTopology();
    lifecycle.onCaughtUp();
    EXPECT_FALSE(lifecycle.isActivated());
    EXPECT_EQ(0, lifecycle.advance());
}

TEST_F(GatewayLifecycleTest, AnActivationSeenWhileReplayingOpensOnlyOnceCaughtUp)
{
    loadTopology();
    lifecycle.onGatewayActive(MY_ID);
    EXPECT_EQ(State::Replaying, lifecycle.state());
    EXPECT_EQ(0, lifecycle.advance());
    EXPECT_EQ(0, actions.count("open"));

    EXPECT_TRUE(lifecycle.onCaughtUp());
    EXPECT_FALSE(lifecycle.onCaughtUp());
    EXPECT_EQ(1, lifecycle.advance());
    EXPECT_EQ(State::Serving, lifecycle.state());
}

TEST_F(GatewayLifecycleTest, StaysShutWhenCaughtUpButNotActivated)
{
    loadTopology();
    lifecycle.onCaughtUp();
    EXPECT_EQ(State::Passive, lifecycle.state());
    EXPECT_EQ(0, lifecycle.advance());
    EXPECT_FALSE(lifecycle.isServing());
    EXPECT_EQ(0, actions.count("open"));
}

TEST_F(GatewayLifecycleTest, PublishesGatewayStartedBeforeOpeningTheGate)
{
    becomeServing();
    EXPECT_TRUE(lifecycle.isServing());
    EXPECT_EQ((std::vector<std::string>{ "identity(5,6)", MY_STARTED, "open" }), actions.calls);
}

TEST_F(GatewayLifecycleTest, DoesNotOpenTheGateWhileGatewayStartedIsBackPressured)
{
    actions.gatewayStartedLands = false;
    becomeServing();
    EXPECT_EQ(State::Passive, lifecycle.state());
    EXPECT_EQ(0, actions.count("open"));

    actions.gatewayStartedLands = true;
    EXPECT_EQ(1, lifecycle.advance());
    EXPECT_EQ(State::Serving, lifecycle.state());
}

TEST_F(GatewayLifecycleTest, RetriesAGateThatDidNotOpenWithoutRegisteringAgain)
{
    actions.gateOpens = false;
    becomeServing();
    EXPECT_EQ(State::Passive, lifecycle.state());

    actions.gateOpens = true;
    EXPECT_EQ(1, lifecycle.advance());
    EXPECT_EQ(State::Serving, lifecycle.state());
    EXPECT_EQ(1, actions.count(MY_STARTED)) << "once per activation, not per attempt";
    EXPECT_EQ(2, actions.count("open"));
}

TEST_F(GatewayLifecycleTest, AGateThatClosesByItselfReopensWithoutRegisteringAgain)
{
    becomeServing();
    lifecycle.onGateClosed();
    EXPECT_EQ(State::Passive, lifecycle.state());
    EXPECT_EQ(0, actions.count("close")) << "the gate is already closed";

    EXPECT_EQ(1, lifecycle.advance());
    EXPECT_EQ(1, actions.count(MY_STARTED));
    EXPECT_EQ(2, actions.count("open"));
}

TEST_F(GatewayLifecycleTest, StandsDownWhenAGatewayActiveNamesTheSibling)
{
    becomeServing();
    lifecycle.onGatewayActive(SIBLING_ID);
    EXPECT_EQ(State::Passive, lifecycle.state());
    EXPECT_FALSE(lifecycle.isActivated());
    EXPECT_EQ(1, actions.count("close"));
    EXPECT_EQ(0, lifecycle.advance());
}

TEST_F(GatewayLifecycleTest, ReopensOnASecondDesignationAsANewEpoch)
{
    becomeServing();
    lifecycle.onGatewayActive(SIBLING_ID);
    lifecycle.onGatewayActive(MY_ID);
    EXPECT_EQ(1, lifecycle.advance());
    EXPECT_EQ(State::Serving, lifecycle.state());
    EXPECT_EQ(2, actions.count(MY_STARTED));
}

TEST_F(GatewayLifecycleTest, IgnoresAnotherLogicalGatewaysElection)
{
    becomeServing();
    lifecycle.onGatewayActive(OTHER_PAIR_ID);
    EXPECT_EQ(State::Serving, lifecycle.state());
    EXPECT_TRUE(lifecycle.isActivated());
    EXPECT_EQ(0, actions.count("close"));
}

TEST_F(GatewayLifecycleTest, IgnoresAnActivationForAnUnknownGateway)
{
    becomeServing();
    lifecycle.onGatewayActive(99);
    EXPECT_EQ(State::Serving, lifecycle.state());
    EXPECT_TRUE(lifecycle.isActivated());
}

TEST_F(GatewayLifecycleTest, ASupersededActivationReplayedOnRestartLeavesItPassive)
{
    loadTopology();
    lifecycle.onGatewayActive(MY_ID);      // history: this instance once served
    lifecycle.onGatewayActive(SIBLING_ID); // and was replaced
    lifecycle.onCaughtUp();
    EXPECT_EQ(0, lifecycle.advance());
    EXPECT_EQ(State::Passive, lifecycle.state());
    EXPECT_EQ(0, actions.count("close")) << "never opened, so nothing to close";
    EXPECT_EQ(0, actions.count(MY_STARTED));
}

TEST_F(GatewayLifecycleTest, RefusesASessionAcquiredWhileTheGateIsShut)
{
    loadTopology();
    lifecycle.onCaughtUp();
    EXPECT_FALSE(lifecycle.onSessionAcquired());

    lifecycle.onGatewayActive(MY_ID);
    lifecycle.advance();
    EXPECT_TRUE(lifecycle.onSessionAcquired());

    lifecycle.onGatewayActive(SIBLING_ID); // the gate closed mid-logon
    EXPECT_FALSE(lifecycle.onSessionAcquired());
}

} // namespace
} // namespace org::limitless::seqeron::app

File: client/src/test/java/org/limitless/seqeron/app/GatewayLifecycleTest.java
package org.limitless.seqeron.app;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.limitless.seqeron.app.GatewayLifecycle.State;

/**
 * Drives {@link GatewayLifecycle} directly; the races it covers are too narrow for an end-to-end script.
 * Merged from phixeron's acceptor and initiator tests. Topology: {@code GW-A} (id 5) and {@code GW-B} (id 6)
 * under {@code gatewaySourceId} 6, beside another pair (ids 1 and 2) under 0.
 *
 * <p>Case for case with {@code GatewayLifecycleTest.cpp}.
 */
class GatewayLifecycleTest {
    private static final String ME = "GW-A";
    private static final int MY_ID = 5;
    private static final int SIBLING_ID = 6;
    private static final int MY_SOURCE_ID = 6;
    private static final int OTHER_PAIR_ID = 1;
    private static final int OTHER_PAIR_SOURCE_ID = 0;

    private static final class RecordingActions implements GatewayLifecycle.Actions {
        private final List<String> calls = new ArrayList<>();
        private boolean gatewayStartedLands = true;
        private boolean gateOpens = true;

        @Override
        public void identityResolved(final int gatewayId, final int gatewaySourceId, final int preferenceRank) {
            calls.add("identity(" + gatewayId + "," + gatewaySourceId + ")");
        }

        @Override
        public boolean publishGatewayStarted(final int gatewayId) {
            calls.add("GatewayStarted(" + gatewayId + ")");
            return gatewayStartedLands;
        }

        @Override
        public boolean openGate() {
            calls.add("open");
            return gateOpens;
        }

        @Override
        public void closeGate() {
            calls.add("close");
        }

        long count(final String call) {
            return calls.stream().filter(call::equals).count();
        }
    }

    private RecordingActions actions;
    private GatewayLifecycle lifecycle;

    @BeforeEach
    void setUp() {
        actions = new RecordingActions();
        lifecycle = new GatewayLifecycle(ME, actions);
    }

    private void loadTopology() {
        lifecycle.onGatewayRegistered(OTHER_PAIR_ID, OTHER_PAIR_SOURCE_ID, "OTHER-A", 0);
        lifecycle.onGatewayRegistered(2, OTHER_PAIR_SOURCE_ID, "OTHER-B", 1);
        lifecycle.onGatewayRegistered(MY_ID, MY_SOURCE_ID, ME, 0);
        lifecycle.onGatewayRegistered(SIBLING_ID, MY_SOURCE_ID, "GW-B", 1);
    }

    private void becomeServing() {
        loadTopology();
        lifecycle.onGatewayActive(MY_ID);
        lifecycle.onCaughtUp();
        lifecycle.advance();
    }

    @Test
    void resolvesItsIdentityFromTheRowNamingIt() {
        loadTopology();
        assertEquals(MY_ID, lifecycle.gatewayId());
        assertEquals(MY_SOURCE_ID, lifecycle.gatewaySourceId());
        assertEquals(List.of("identity(" + MY_ID + "," + MY_SOURCE_ID + ")"), actions.calls);
    }

    @Test
    void ignoresAnActivationBeforeAnyRowNamesThisInstance() {
        lifecycle.onGatewayActive(MY_ID);
        loadTopology();
        lifecycle.onCaughtUp();
        assertFalse(lifecycle.isActivated());
        assertEquals(0, lifecycle.advance());
    }

    @Test
    void anActivationSeenWhileReplayingOpensOnlyOnceCaughtUp() {
        loadTopology();
        lifecycle.onGatewayActive(MY_ID);
        assertEquals(State.REPLAYING, lifecycle.state());
        assertEquals(0, lifecycle.advance());
        assertEquals(0, actions.count("open"));

        assertTrue(lifecycle.onCaughtUp());
        assertFalse(lifecycle.onCaughtUp());
        assertEquals(1, lifecycle.advance());
        assertEquals(State.SERVING, lifecycle.state());
    }

    @Test
    void staysShutWhenCaughtUpButNotActivated() {
        loadTopology();
        lifecycle.onCaughtUp();
        assertEquals(State.PASSIVE, lifecycle.state());
        assertEquals(0, lifecycle.advance());
        assertFalse(lifecycle.isServing());
        assertEquals(0, actions.count("open"));
    }

    @Test
    void publishesGatewayStartedBeforeOpeningTheGate() {
        becomeServing();
        assertTrue(lifecycle.isServing());
        assertEquals(List.of("GatewayStarted(" + MY_ID + ")", "open"), actions.calls.subList(1, 3));
    }

    @Test
    void doesNotOpenTheGateWhileGatewayStartedIsBackPressured() {
        actions.gatewayStartedLands = false;
        becomeServing();
        assertEquals(State.PASSIVE, lifecycle.state());
        assertEquals(0, actions.count("open"));

        actions.gatewayStartedLands = true;
        assertEquals(1, lifecycle.advance());
        assertEquals(State.SERVING, lifecycle.state());
    }

    @Test
    void retriesAGateThatDidNotOpenWithoutRegisteringAgain() {
        actions.gateOpens = false;
        becomeServing();
        assertEquals(State.PASSIVE, lifecycle.state());

        actions.gateOpens = true;
        assertEquals(1, lifecycle.advance());
        assertEquals(State.SERVING, lifecycle.state());
        assertEquals(1, actions.count("GatewayStarted(" + MY_ID + ")"), "once per activation, not per attempt");
        assertEquals(2, actions.count("open"));
    }

    @Test
    void aGateThatClosesByItselfReopensWithoutRegisteringAgain() {
        becomeServing();
        lifecycle.onGateClosed();
        assertEquals(State.PASSIVE, lifecycle.state());
        assertEquals(0, actions.count("close"), "the gate is already closed");

        assertEquals(1, lifecycle.advance());
        assertEquals(1, actions.count("GatewayStarted(" + MY_ID + ")"));
        assertEquals(2, actions.count("open"));
    }

    @Test
    void standsDownWhenAGatewayActiveNamesTheSibling() {
        becomeServing();
        lifecycle.onGatewayActive(SIBLING_ID);
        assertEquals(State.PASSIVE, lifecycle.state());
        assertFalse(lifecycle.isActivated());
        assertEquals(1, actions.count("close"));
        assertEquals(0, lifecycle.advance());
    }

    @Test
    void reopensOnASecondDesignationAsANewEpoch() {
        becomeServing();
        lifecycle.onGatewayActive(SIBLING_ID);
        lifecycle.onGatewayActive(MY_ID);
        assertEquals(1, lifecycle.advance());
        assertEquals(State.SERVING, lifecycle.state());
        assertEquals(2, actions.count("GatewayStarted(" + MY_ID + ")"));
    }

    @Test
    void ignoresAnotherLogicalGatewaysElection() {
        becomeServing();
        lifecycle.onGatewayActive(OTHER_PAIR_ID);
        assertEquals(State.SERVING, lifecycle.state());
        assertTrue(lifecycle.isActivated());
        assertEquals(0, actions.count("close"));
    }

    @Test
    void ignoresAnActivationForAnUnknownGateway() {
        becomeServing();
        lifecycle.onGatewayActive(99);
        assertEquals(State.SERVING, lifecycle.state());
        assertTrue(lifecycle.isActivated());
    }

    @Test
    void aSupersededActivationReplayedOnRestartLeavesItPassive() {
        loadTopology();
        lifecycle.onGatewayActive(MY_ID); // history: this instance once served
        lifecycle.onGatewayActive(SIBLING_ID); // and was replaced
        lifecycle.onCaughtUp();
        assertEquals(0, lifecycle.advance());
        assertEquals(State.PASSIVE, lifecycle.state());
        assertEquals(0, actions.count("close"), "never opened, so nothing to close");
        assertEquals(0, actions.count("GatewayStarted(" + MY_ID + ")"));
    }

    @Test
    void refusesASessionAcquiredWhileTheGateIsShut() {
        loadTopology();
        lifecycle.onCaughtUp();
        assertFalse(lifecycle.onSessionAcquired());

        lifecycle.onGatewayActive(MY_ID);
        lifecycle.advance();
        assertTrue(lifecycle.onSessionAcquired());

        lifecycle.onGatewayActive(SIBLING_ID); // the gate closed mid-logon
        assertFalse(lifecycle.onSessionAcquired());
    }
}
